File: packages/sdk-ios/Sources/QSDK/Transport/ViewSnapshot.swift
// Copyright (C) 2026 Ning Zou <q.cue.2026@gmail.com>
// SPDX-License-Identifier: AGPL-3.0-only

import UIKit

/// Captures the view hierarchy as a JSON string for the Q backend.
/// Privacy-filtered: no text content, skips secure/sensitive fields.
struct ViewSnapshot {

    struct ViewNode: Codable {
        let cls: String              // e.g. "UIButton"
        let id: String?              // accessibilityIdentifier
        let label: String?           // accessibilityLabel
        let frame: [Int]             // [x, y, width, height]
        let isAccessible: Bool
        let children: [ViewNode]?
    }

    /// Maximum number of nodes to capture.
    private static let maxNodes = 200

    /// Sensitive identifier patterns (case-insensitive).
    private static let sensitivePatterns = ["password", "cvv", "ssn", "pin", "secret"]

    /// Capture the current key window's view hierarchy as a JSON string.
    static func capture() -> String? {
        guard let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = scene.windows.first(where: { $0.isKeyWindow }) else {
            return nil
        }

        var count = 0
        let root = walk(view: window, count: &count)
        guard let data = try? JSONEncoder().encode(root) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private static func walk(view: UIView, count: inout Int) -> ViewNode? {
        guard count < maxNodes else { return nil }

        // Skip secure text entries
        if let textField = view as? UITextField, textField.isSecureTextEntry { return nil }

        // Skip views with sensitive identifiers
        let allIds = [view.accessibilityIdentifier, view.accessibilityLabel].compactMap { $0?.lowercased() }
        for id in allIds {
            for pattern in sensitivePatterns {
                if id.contains(pattern) { return nil }
            }
        }

        count += 1

        let frame = view.frame
        let children: [ViewNode]? = view.subviews.isEmpty ? nil : view.subviews.compactMap { walk(view: $0, count: &count) }

        return ViewNode(
            cls: String(describing: type(of: view)),
            id: view.accessibilityIdentifier,
            label: view.accessibilityLabel,
            frame: [Int(frame.origin.x), Int(frame.origin.y), Int(frame.width), Int(frame.height)],
            isAccessible: view.isAccessibilityElement,
            children: children?.isEmpty == true ? nil : children
        )
    }
}

File: packages/sdk-android/src/main/kotlin/dev/useq/sdk/transport/ViewSnapshot.kt
// Copyright (C) 2026 Ning Zou <q.cue.2026@gmail.com>
// SPDX-License-Identifier: AGPL-3.0-only

package dev.useq.sdk.transport

import android.app.Activity
import android.text.InputType
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import com.google.gson.Gson

/**
 * Captures the view hierarchy as a JSON string for the Q backend.
 * Privacy-filtered: no text content, skips secure/sensitive fields.
 *
 * Primary approach: traverse Activity.window.decorView.rootView recursively.
 * For Compose views, attempts to access SemanticsNode tree (graceful fallback if unavailable).
 */
object ViewSnapshot {

    data class ViewNode(
        val cls: String,
        val id: String? = null,
        val desc: String? = null,
        val bounds: List<Int>,
        val clickable: Boolean = false,
        val children: List<ViewNode>? = null,
    )

    private const val MAX_NODES = 200
    private val sensitivePatterns = listOf("password", "cvv", "ssn", "pin", "secret")
    private val gson = Gson()

    /**
     * Capture the current activity's view hierarchy as a JSON string.
     */
    fun capture(activity: Activity): String? {
        val rootView = activity.window?.decorView?.rootView ?: return null
        var count = 0
        val root = walk(activity, rootView, count = { count++ ; count })
        return try { gson.toJson(root) } catch (_: Exception) { null }
    }

    private fun walk(activity: Activity, view: View, count: () -> Int): ViewNode? {
        if (count() > MAX_NODES) return null

        // Skip password fields
        if (view is EditText) {
            val inputType = view.inputType
            if (inputType and InputType.TYPE_TEXT_VARIATION_PASSWORD != 0 ||
                inputType and InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD != 0) {
                return null
            }
        }

        // Skip views with sensitive identifiers
        val resName = try {
            if (view.id != View.NO_ID) activity.resources.getResourceEntryName(view.id) else null
        } catch (_: Exception) { null }

        val contentDesc = view.contentDescription?.toString()

        val allIds = listOfNotNull(resName?.lowercase(), contentDesc?.lowercase())
        for (id in allIds) {
            for (pattern in sensitivePatterns) {
                if (id.contains(pattern)) return null
            }
        }

        // Bounds
        val rect = android.graphics.Rect()
        view.getGlobalVisibleRect(rect)
        val bounds = listOf(rect.left, rect.top, rect.width(), rect.height())

        // Children
        val children = if (view is ViewGroup && view.childCount > 0) {
            (0 until view.childCount).mapNotNull { i ->
                walk(activity, view.getChildAt(i), count)
            }.ifEmpty { null }
        } else null

        return ViewNode(
            cls = view.javaClass.simpleName,
            id = resName,
            desc = contentDesc,
            bounds = bounds,
            clickable = view.isClickable,
            children = children,
        )
    }
}
